### ISSUE TITLE
Add support for native python jitclass @staticmethod

### ISSUE THREAD

**Issue by stuz5000** [numba/numba#4451]
It would be nice to have support for classmethods/staticmethods within a jitclass.

Here's a simple study. I'd like to add a `.transpose` property to invert my Graph class:
```
import numba
import numpy as np

@numba.jitclass(
    [("parent", numba.int32[:]),  
     ("child", numba.int32[:])]
)
class Graph(object):
    def __init__(self, parent, child):
        self.parent = parent
        self.child = child

    # Attempt 1. Can't return new object from jitclass function
    #@property
    #def transpose(self):
     #    """ Flip parent and child """
     #    with numba.objmode:
     #        return Graph(self.child, self.parent)

    # Attempt 2. TypeError: class members are not yet supported: transpose
    # @staticmethod
    # def transpose(obj):
    #     """ Flip parent and child """
    #     return Graph(obj.child, obj.parent)


# Attempt 1.5. Can't add object mode stuff in subclasses.
class Graph2(Graph):
    @property
    def transpose(self): pass

# Attempt 3.
def Graph_transpose(obj):
    """ Flip parent and child """
    return Graph(obj.child, obj.parent)

g = Graph(np.zeros(100, dtype=np.int32),np.zeros(100, dtype=np.int32))
# h = g.transpose     #  property
# h = g.transpose()  # staticmethod
h = Graph_transpose(g) # what's currently possible
```

... so I don't think its possible to 'encapsulate' utility methods within the class itself. This makes navigating the codebase difficult and leads to people implementing the same functions over and over because there's no obvious way to attach an objmode function to a jitclass -- even as a staticmethod. 

I suspect it will be a while before we can return new instances from jitclass methods, so why not allows @staticmethods? Since there's no need for these to be compiled (and if necessary we can add @jit), why not allow python staticmethds in jitclasses?

**Comment by DiSchi123** [comment]
I too would like this to be added.

**Comment by EPronovost** [comment]
I [started work on this](https://github.com/EPronovost/numba/tree/epronovost/jitclass-staticmethod).  What I have works for this 
```
@jitclass([("x", intp)])
class Test:
    def __init__(self, x):
        self.x = x
        
    @staticmethod
    def add(x, y):
        return x + y

t = Test(4)
Test.add(1, 2)
```
but `t.add(1, 2)` fails with `AttributeError: 'Test' object has no attribute 'add'`

**Comment by EPronovost** [comment]
I believe we can close this?

**Comment by DiSchi123** [comment]
That's great! I have not tested it personally yet, which release is this?

**Comment by EPronovost** [comment]
Static methods for `jitclass` are part of the latest release.

**Comment by stuartarchibald** [comment]
> I believe we can close this?

Agree. Thanks for implementing.

**Comment by calbaker** [comment]
@stuartarchibald , should this be expected to work?

```
@dataclass
class MyDataClass(object):
    a: int = 1
    b: int = 2
    c: int = 3
    
myinst = MyDataClass(1, 2, 3)
    
@jitclass([
    ('a', int64),
    ('b', int64),
    ('c', int64),    
])
class MyJitClass(MyDataClass):
    pass
    def __hash__(self):
        pass
    def __dataclass_params__(self):
        pass
    def __dataclass_fields__(self):
        pass
    
myjitinst = MyJitClass(1, 2, 3)
```

It fails with `TypeError: class members are not yet supported: a, b, c`

**Comment by calbaker** [comment]
if I don't provide default values in the `dataclass`, it works

**Comment by stuartarchibald** [comment]
@calbaker I don't think `@dataclass` has been tried out/tested in this context. If you discover useful things please feel free to open an issue to discuss or PR regarding them against the docs/unittests. Thanks!
